### DISCUSSION TITLE
Ability to chain tasks like deno-task-shell

### DISCUSSION THREAD

**Discussion by Hexagon** [Hexagon/pup#19] [Ideas]
Ability to chain processes using && and || just like https://github.com/denoland/deno_task_shell

Would allow for some cool solutions where only the last result of the chain affects restart strategies etc.

Example configuration where task 2 only runs if task1 fals
{
   processes [
      {
         "id": "my-task-chain",
         "cmd": ["deno","run","task1.ts","||","deno","run","task2.ts"],
         "autostart": true
      }
   } 
}

Thoughts?

Other uses for running all commands through a custom shell?

**Comment by Leokuma** [comment]
Not sure if possible, but what if we use [dax](https://github.com/dsherret/dax) as the default interpreter for every cmd? I guess it's just a wrapper around `deno_task_shell`. It even has a [cross platform `unset`](https://github.com/dsherret/dax#custom-cross-platform-shell-commands) that could be used to filter env variables.

Sorry I've been so busy lately, but still following the project 👍.

**Comment by Hexagon** [comment]
I can make a try using dax, the code to run processes is only about 20 sloc 😃 

https://github.com/Hexagon/pup/blob/6dbac5efa95bc61fc4cab3096d15a0107911a94c/lib/core/runner.ts#L80-L103

@dsherret I'm working on a poc dax runner. Checking out the documentation of dax, a couple of (hopefully) quick questions pops up that would be great if you could have a look at 🙏 :

```ts
let child = $.raw`${commandString}`.stdout("piped").stderr("piped") <-
                          // Question 1. Is it possible to have a fully dynamic command string with 
                          // arguments, like ["deno run -A","/path/with spaces and/stuff",
                          // "&& echo","Hello world!"]

if (this.processConfig.cwd) child = child.cwd(this.processConfig.cwd)
if (env) child = child.env(env)
this.process = child.spawn()
this.process = child.spawn()
runningCallback(/*pid*/0) // <- Question 2: How to get the pid of the process, is it possible 
                          //    today, is it even relevant when running processes using a shell? 
                          //    Optimally i would want the pid of the currently executing process if many are chained
this.writePidFile()
this.pipeToLogger("stdout", this.process.stdout())
this.pipeToLogger("stderr", this.process.stderr())
const result = await this.process
// this.process.ref()     // <- Question 3: Is it possible to ref the child process to force it exit 
                          //    along with the parent using dax?
```

And the last question, are there any obvious reasons why i get "Pipe error" every now and then?

![image](https://github.com/Hexagon/pup/assets/419737/767e5ebc-f800-4598-94a2-416a6021d08c)

**Comment by Hexagon** [comment]
~~POC pushed to branch [feature/dax-shell](https://github.com/Hexagon/pup/tree/feature/dax-shell)~~

Implemented in [main](https://github.com/Hexagon/pup/tree/main) and included in pup since `1.0.0-beta.23`

https://github.com/Hexagon/pup/blob/21fd1bd6a921176d60ffe99acbed88653078796e/lib/core/runner.ts#L87
